network/gossip/libp2p/peerManager.go

```diff
@@ -43,56 +44,43 @@ type PeerManager struct {
 
 // NewPeerManager creates a new peer manager which calls the idsProvider callback to get a list of peers to connect to
 // and it uses the connector to actually connect or disconnect from peers.
-func NewPeerManager(ctx context.Context, logger zerolog.Logger, idsProvider func() (flow.IdentityList, error), connector Connector) *PeerManager {
+func NewPeerManager(ctx context.Context, logger zerolog.Logger, idsProvider func() (flow.IdentityList, error),
+	connector Connector) *PeerManager {
 	return &PeerManager{
 		ctx:          ctx,
+		unit:         engine.NewUnit(),
 		logger:       logger,
 		idsProvider:  idsProvider,
 		connector:    connector,
 		peerRequestQ: make(chan struct{}, 1),
 	}
 }
 
-// Start kicks off the ambient periodic connection updates
-// It is blocked till the connection update is kicked off
-func (pm *PeerManager) Start() error {
-	wg := &sync.WaitGroup{}
-	wg.Add(2)
+// Ready kicks off the ambient periodic connection updates.
+func (pm *PeerManager) Ready() <-chan struct{} {
+	// makes sure that peer update request is invoked
+	// once before returning
+	pm.RequestPeerUpdate()
+
+	// also starts running it periodically
+	pm.unit.LaunchPeriodically(pm.RequestPeerUpdate, PeerUpdateInterval, time.Duration(0))
 
-	go pm.updateLoop(wg)
-	go pm.periodicUpdate(wg)
+	pm.unit.Launch(pm.updateLoop)
+
+	return pm.unit.Ready()
+}
 
-	wg.Wait()
-	return nil
+func (pm *PeerManager) Done() <-chan struct{} {
+	return pm.unit.Done()
 }
 
 // updateLoop triggers an update peer request when it has been requested
-func (pm *PeerManager) updateLoop(wg *sync.WaitGroup) {
-	wg.Done()
+func (pm *PeerManager) updateLoop() {
 	for {
 		select {
-		case <-pm.ctx.Done():
-			return
 		case <-pm.peerRequestQ:
 			pm.updatePeers()
-		}
-	}
-}
-
-// updateLoop request periodic connection update
-func (pm *PeerManager) periodicUpdate(wg *sync.WaitGroup) {
-	wg.Done()
-	// request initial discovery
-	pm.RequestPeerUpdate()
-
-	ticker := time.NewTicker(PeerUpdateInterval)
-	defer ticker.Stop()
-
-	for {
-		select {
-		case <-ticker.C:
-			pm.RequestPeerUpdate()
-		case <-pm.ctx.Done():
+		case <-pm.unit.Ctx().Done():
```

Comment:
Better to use `unit.Quit()` here. The fact that it uses a context under the covers is more of an implementation detail IMO.

Reply:
Fixed in https://github.com/onflow/flow-go/pull/99/commits/14e5a0bcf79ed31b89f81d442e27954d13b256af